Fix scheduler_perf test so that QPS is non-zero even if there is a scheduling "cold start"

jayunit100 opened the pull request:
@gmarek  ... is something like this more realistic as an expectation ?  That is, wait till scheduling starts, then wait 1 second before any polling attempts....  i.e. 

- gaurantee uniform sleep before measure, by doing it first rather then last
- print the initial status so that its easier to debug in case of issues

#36532

<!-- Reviewable:start -->

---
This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/36543)
<!-- Reviewable:end -->

jayunit100 committed:
- Fix scheduler_perf test so that QPS are not zero due to cold start zero
- Wait for 1% of scheduling to start before measuring QPS w map

jayunit100 commented:
not sure if the other 0 based failures might happen because of the polling list, maybe a watch would be more stable? not sure.

jayunit100 commented:
In any case, you can still see 0 scheduling events, even after the warm up.
```
4s      rate: [6 0 799 1071]    total: 1876
5s      rate: [6 0 799 1071 1030]       total: 2906
6s      rate: [6 0 799 1071 1030 94]    total: 3000
Scheduled 3000 Pods in 6 seconds (500 per second on average). min QPS was 6
```
Is this a bug? Or are we overtesting (i.e. do we really expect the scheduler to always be putting pods down, every second?)

gmarek commented on test/integration/scheduler_perf/scheduler_test.go:
```diff
@@ -194,26 +194,45 @@ func schedulePods(config *testConfig) int32 {
 		prev = len(scheduled)
-		time.Sleep(1 * time.Second)
+
+		// This can potentially affect performance of scheduler, since List() is done under mutex.
```
Well, not this - the List in line 217 can.

gmarek commented on test/integration/scheduler_perf/scheduler_test.go:
```diff
@@ -194,26 +194,45 @@ func schedulePods(config *testConfig) int32 {
-		// This can potentially affect performance of scheduler, since List() is done under mutex.
-		// Listing 10000 pods is an expensive operation, so running it frequently may impact scheduler.
-		// TODO: Setup watch on apiserver and wait until all pods scheduled.
+		time.Sleep(1 * time.Second)
```
Is there a reason why you moved the seep here?

gmarek commented on test/integration/scheduler_perf/scheduler_test.go:
```diff
@@ -194,26 +194,45 @@ func schedulePods(config *testConfig) int32 {
 	start := time.Now()
+
+	// Bake in time for the first pod scheduling event.
+	waitForStart := func() {
```
Why you need this to be a function? Can't you just inline this (with a slight difference that you want to `break` instead of `return`)?

jayunit100 replied:
just for readability.  IMO `waitForStart` in one line proceeding the "actual" test is instantly readable even if you've never seen the code before.

gmarek replied:
Yeah, but defining the function and calling it straight after that doesn't improve readability at all. Please just inline.

jayunit100 commented:
@gmarek agree the last artifact pod set isnt useful good catch

jayunit100 commented:
Ok @gmarek minimized the changes.  if the inner `waitForStart` function is a problem i can get rid of it, i just think it reads more cleanly.

- reverted the earlier return: you're right it will skew the QPS results !
- also reverted the `time.Sleep` at the beggining.  Its just an extra way to gaurd against cold starts but i think its unnecessary

gmarek commented on test/integration/scheduler_perf/scheduler_test.go:
```diff
@@ -194,24 +194,42 @@ func schedulePods(config *testConfig) int32 {
+
+	// Now that scheduling has started, lets start taking the pulse on how many pods are happening per second.
 	for {
+		scheduled := config.schedulerConfigFactory.ScheduledPodLister.Indexer.List()
```
This should be under the comment that refers to it.

jayunit100 commented:
as is I think this incrementally makes this test more stable.  `TestSchedule1000Node30KPods` is still unstable i think, but others pass once cold start failures are avoided.

gmarek commented:
1000 Node one may timeout... I'm not sure what's the current policy around merges, so I'm just applying `lgtm` label. I don't know if we merge things like this now...

jayunit100 commented:
updated to fix the cold start once and for all..... :) thanks!

jayunit100 commented on test/integration/scheduler_perf/scheduler_test.go:
```diff
@@ -194,25 +194,47 @@ func schedulePods(config *testConfig) int32 {
+
 		// There's no point in printing it for the last iteration, as the value is random
 		qps := len(scheduled) - prev
+		qpsStats[qps] += 1
```
this also adds in a map,  useful for debugging on different situations i.e. on VMs where there may be lots of zeros...   total length of the map is relatively small since it aggregates, like 10 keys or so.

gmarek commented on test/integration/scheduler_perf/scheduler_test.go:
```diff
@@ -194,25 +194,47 @@ func schedulePods(config *testConfig) int32 {
+			break
+		}
+	}
+	// map minimum QPS entries in a counter.  useful for debugging tests.
```
double space after period.

gmarek commented on test/integration/scheduler_perf/scheduler_test.go:
```diff
@@ -194,25 +194,47 @@ func schedulePods(config *testConfig) int32 {
 		}
-
-		fmt.Printf("%ds\trate: %d\ttotal: %d\n", time.Since(start)/time.Second, qps, len(scheduled))
+		fmt.Printf("%ds\trate: %d\ttotal: %d (qps frequency: %v)\n", time.Since(start)/time.Second, qps, len(scheduled), qpsStats)
```
I'd print qps stat at the end of the test, not in every iteration.

jayunit100 replied:
the nice thing about the qps every iteration is that you can "see" a test fail.

gmarek commented on test/integration/scheduler_perf/scheduler_test.go:
```diff
@@ -194,25 +194,47 @@ func schedulePods(config *testConfig) int32 {
+		scheduled := config.schedulerConfigFactory.ScheduledPodLister.Indexer.List()
+		// 30,000 pods -> wait till @ least 300 are scheduled to start measuring.
+		// TODO Find out why sometimes there may be scheduling blips in the beggining.
+		if len(scheduled) > config.numPods/100 {
```
I don't understand, but I also don't care enough:) It seems that it's some problem with a test setup, but if this makes results stable I'm completely fine with that.

jayunit100 replied:
agree, its a very weird issue.

timothysc replied:
So here is my take on some of this... I think the algorithm for evaluation here is flawed.   I think we should bin, throw out the high and low and take the average, as lumps should be expected.  This jiggery is far to brittle for my liking.  Standard pattern should be to normalize the curves and eliminate the outliers to determine base level drift on the average distribution.

gmarek commented:
Few nits

timothysc commented:
@k8s-bot kops aws e2e test this

timothysc commented:
@k8s-bot kops aws e2e test this

timothysc commented:
@k8s-bot test this #IGNORE

timothysc commented:
I don't know why kops restart fails on this PR, or why it's blocking it from merge either.  
@k8s-oncall

krousey commented:
@zmerlynn Any ideas what is up with the AWS PR builder?
